### PR TITLE
Fixed point and signed in `dat`.

### PR DESCRIPTION
**Pull request by cgyurgyik**
This is heavily based off @yoonachang 's PR #292. I didn't branch off this simply because the number of merge conflicts was a pain. She did most of the work, I am simply adding a bit of clean-up and a few small syntactical changes.

Co-authored with @yoonachang.
Closes #292.

### CODE REVIEW

**Comment by cgyurgyik** [comment]
@rachitnigam When you have a moment, can you take a quick glance at the `.dat` JSON format? I'll need to eventually refactor all test files, so I want to make sure this is nailed down.

**Comment by rachitnigam** [comment]
I would put the format details under a new dictionary called “format” so that each memory specifies a “data” field and a “format” field.

Also, for fixed point, how do you feel about having a “fraction_width” field too? It’s redundant but it supports consumers that might want to use <int_width,frac_width>.

**Comment by cgyurgyik** [comment]
> I would put the format details under a new dictionary called “format” so that each memory specifies a “data” field and a “format” field.
> 
> Also, for fixed point, how do you feel about having a “fraction_width” field too? It’s redundant but it supports consumers that might want to use <int_width,frac_width>.

Done. A new format dictionary is used. Users can now provide any of the following pairs: 
1. <int, frac>
2. <width, int>
3. <width, frac> 

since the third can always be inferred. Let me know what you think.

**Comment by rachitnigam** [comment]
Yeah, this looks good!

**Comment by cgyurgyik** [comment]
Ok, so this PR adds parsing for (un/signed) fixed point and (un/signed) bitnum. I've added a few small tests to verify it works correctly. However, they aren't thorough. 

The next step is to test each primitive carefully, verifying the correctness of the `.dat` parser and the actual primitives.

Currently, when a number cannot be represented in fixed point, it errors out. We can eventually relax this condition.

The testing is in `tests/correctness/numeric-types`. There are 3 sub-directories:
1. `parsing`: Basic tests to ensure that formats are parsed correctly. Perhaps not necessary, however useful during the build process.
2. `fixed-point`: All fixed point related tests.
3. `bitnum`: All bitnum related tests.

File: tests/correctness/numeric-types/fixed-point/diff-width-add.expect [hunk 1]
```diff
@@ -1,39 +1,39 @@
 {
   "mem0": [
     [
-      12,
+      1.5,
```
Comment (rachitnigam) [inline tests/correctness/numeric-types/fixed-point/diff-width-add.expect:4]: This test is weird. Why are the remaining three values all zeros?

File: tests/correctness/numeric-types/fixed-point/diff-width-add.futil.data [hunk 2]
```diff
@@ -0,0 +1,44 @@
+{
+  "mem0": {
+    "data": [
+        [1.5,0,0,0],
```
Comment (rachitnigam) [inline tests/correctness/numeric-types/fixed-point/diff-width-add.futil.data:4]: Same question as above—these zeros should probably be actual values. Good opportunity to document overflow behavior.

File: tests/correctness/relay/relu.expect [hunk 3]
```diff
@@ -1,7 +1,7 @@
 {
   "x": [
-    4294967295,
-    4294967294,
+    -1,
```
Comment (rachitnigam) [inline tests/correctness/relay/relu.expect:3]: Sweet!

File: tests/correctness/udot_product.futil [hunk 4]
```diff
@@ -4,9 +4,9 @@ import "primitives/bitnum/signed.futil";
 component main() -> () {
```
Comment (rachitnigam) [inline tests/correctness/udot_product.futil:4]: Maybe rename the test file to `unsigned-dot-product.futil`?

**Comment by rachitnigam** [comment]
Minor comments. Squash and merge after fixing. Also, presumably this subsumes #292? If so, you can say Closes #292 in the description and github will automatically close that PR when this is merged.

File: tests/correctness/numeric-types/fixed-point/diff-width-add.futil.data (hunk 2 above)
Comment (cgyurgyik) [reply tests/correctness/numeric-types/fixed-point/diff-width-add.futil.data]: Agreed, I copied this over from Yoona's PR. I'm going to remove it for now. I want to test the primitives in a more systematic way, i.e. For each numeric type, make a list of corner cases and our expectations. I plan to write this up.

File: tests/correctness/udot_product.futil (hunk 4 above)
Comment (cgyurgyik) [reply tests/correctness/udot_product.futil:4]: Done.
